contract/models/contract.py

```diff
@@ -595,23 +599,47 @@ def _recurring_create_invoice(self, date_ref=False):
         return moves
 
     @api.model
-    def cron_recurring_create_invoice(self, date_ref=None):
+    def _get_recurring_create_func(self, create_type="invoice"):
+        """
+        Allows to retrieve the recurring create function depending
+        on generate_type attribute
+        """
+        if create_type == "invoice":
+            return self.__class__._recurring_create_invoice
+
+    @api.model
+    def _cron_recurring_create(self, date_ref=False, create_type="invoice"):
+        """
+        The cron function in order to create recurrent documents
+        from contracts.
+        """
+        _recurring_create_func = self._get_recurring_create_func(
+            create_type=create_type
+        )
         if not date_ref:
             date_ref = fields.Date.context_today(self)
         domain = self._get_contracts_to_invoice_domain(date_ref)
-        invoices = self.env["account.move"]
+        domain = expression.AND(
+            [
+                domain,
+                [("generation_type", "=", create_type)],
+            ]
+        )
+        contracts = self.search(domain)
+        companies = set(contracts.mapped("company_id"))
         # Invoice by companies, so assignation emails get correct context
-        companies_to_invoice = self.read_group(domain, ["company_id"], ["company_id"])
-        for row in companies_to_invoice:
-            contracts_to_invoice = (
-                self.search(row["__domain"])
-                .with_context(allowed_company_ids=[row["company_id"][0]])
-                .filtered(
-                    lambda a: not a.date_end or a.recurring_next_date <= a.date_end
-                )
-            )
-            invoices |= contracts_to_invoice._recurring_create_invoice(date_ref)
-        return invoices
+        # Use ids instead of recordset for perfomances reasons
```

Comment:
I think this comment doesn't apply anymore, isn't it?

Reply:
Done